lib/buffer.js

```diff
@@ -557,21 +557,39 @@ Buffer.prototype.compare = function compare(target,
 // - encoding - an optional encoding, relevant is val is a string
 // - dir - true for indexOf, false for lastIndexOf
 function bidirectionalIndexOf(buffer, val, byteOffset, encoding, dir) {
-  if (typeof byteOffset === 'string') {
-    encoding = byteOffset;
-    byteOffset = undefined;
-  } else if (byteOffset > 0x7fffffff) {
-    byteOffset = 0x7fffffff;
-  } else if (byteOffset < -0x80000000) {
-    byteOffset = -0x80000000;
-  }
-  byteOffset = +byteOffset;  // Coerce to Number.
-  if (isNaN(byteOffset)) {
-    // If the offset is undefined, null, NaN, "foo", etc, search whole buffer.
-    byteOffset = dir ? 0 : (buffer.length - 1);
-  }
+
   dir = !!dir;  // Cast to bool.
 
+  if (typeof byteOffset === 'number') {
+    // Make the general case fast: byteOffset is a number
+    if (byteOffset >= -0x80000000 && byteOffset <= 0x7fffffff) {
+	//general case
+    } else if (byteOffset > 0x7fffffff) {
+      byteOffset = 0x7fffffff;
+    } else if (byteOffset < -0x80000000) {
+      byteOffset = -0x80000000;
+    } else {
+      // No domain value possible remained except NaN
+      byteOffset = dir ? 0 : (buffer.length - 1);
+    }
+  } else {
+    if (typeof byteOffset === 'string') {
```

Comment:
Hello, 
I'm not really sure about this line, I feel like the whole if should be removed. 
I'm not even sure that my understanding of this condition is correctly interpreted, if it were up to me, I'd drop the whole if, but I have what kind of dependency might break as a result.